nitransforms/interp/bspline.py

```diff
@@ -0,0 +1,95 @@
+# emacs: -*- mode: python-mode; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the NiBabel package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Interpolate with 3D tensor-product B-Spline basis."""
+import numpy as np
+import nibabel as nb
+from scipy.sparse import csr_matrix, kron
+
+
+def _cubic_bspline(d, order=3):
+    """Evaluate the cubic bspline at distance d from the center."""
+    if order != 3:
+        raise NotImplementedError
+
+    return np.piecewise(
+        d,
+        [d < 1.0, d >= 1.0],
+        [
+            lambda d: (4.0 - 6.0 * d ** 2 + 3.0 * d ** 3) / 6.0,
+            lambda d: (2.0 - d) ** 3 / 6.0,
+        ],
+    )
+
+
+def grid_bspline_weights(target_nii, ctrl_nii):
+    r"""
+    Evaluate tensor-product B-Spline weights on a grid.
+
+    For each of the *N* input samples :math:`(s_1, s_2, s_3)` and *K* control
+    points or *knots* :math:`\mathbf{k} =(k_1, k_2, k_3)`, the tensor-product
```

Comment:
This math seems off. In the equation, you have `s_i - k_i`, but nothing here constrains `N = K`.

Reply:
*i* indexes the dimension (therefore 1, 2, 3 subscripts). *N* is the number of voxels (reference.dataobj.size) and *K* the number of control points (knots.dataobj.size). I have deeply re-written the equation to make it clearer.